### PR TITLE
add generic auth method

### PR DESCRIPTION
**Pull request by JiahuiZhao11**
Ⅰ. Describe what this PR does
1. add generic auth method
2. modify auth middleware logic
3. fix read audit service address
Ⅱ. Does this pull request fix one issue?
Ⅲ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.
Ⅳ. Describe how to verify it
Ⅴ. Special notes for reviews

### CODE REVIEW

File: cmd/cube/app/options/options.go [hunk 1]
```diff
@@ -18,11 +18,11 @@ package options
 
 import (
 	"fmt"
+	"github.com/kubecube-io/kubecube/pkg/authentication"
```
Comment (weilaaa) [inline cmd/cube/app/options/options.go:21]: The import of the package needs to follow the specification

File: pkg/authentication/identityprovider/generic/generic.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	"github.com/kubecube-io/kubecube/pkg/authentication/identityprovider"
+	"github.com/kubecube-io/kubecube/pkg/clog"
+	"io/ioutil"
+	"net/http"
+)
+
+var Config = authentication.GenericConfig{}
+
+type HeaderProvider struct {
+	URL    string `json:"url,omitempty" yaml:"url"`
```
Comment (weilaaa) [inline pkg/authentication/identityprovider/generic/generic.go:30]: why  do we need yaml tag

Comment (weilaaa) [inline pkg/authentication/identityprovider/generic/generic.go:30]: why  do we need yaml tag

File: pkg/authentication/identityprovider/generic/generic.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+type GenericIdentity struct {
+	Username string
+	Header   http.Header
+}
+
+func (g *GenericIdentity) GetRespHeader() http.Header {
+	return g.Header
+}
+
+func (g *GenericIdentity) GetUserEmail() string {
```
Comment (weilaaa) [inline pkg/authentication/identityprovider/generic/generic.go:43]: add comment if generic has no email

File: pkg/authentication/identityprovider/generic/generic.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		clog.Error("read response error: %v", err)
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		clog.Error("response from the third party is not ok, response is %s", string(respBody))
+		return nil, nil
```
Comment (weilaaa) [inline pkg/authentication/identityprovider/generic/generic.go:90]: return nil error here?

File: pkg/authentication/identityprovider/generic/generic.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	var respMap map[string]interface{}
+	if err = json.Unmarshal(respBody, &respMap); err != nil {
+		clog.Error("json unmarshal error: %v", err)
+		return nil, err
+	}
+
+	username := respMap["name"]
```
Comment (weilaaa) [inline pkg/authentication/identityprovider/generic/generic.go:99]: what if response body has no that key "name"?

File: pkg/authentication/identityprovider/generic/generic.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	respHeader := resp.Header
+
+	return &GenericIdentity{
+		Username: username.(string),
```
Comment (weilaaa) [inline pkg/authentication/identityprovider/generic/generic.go:103]: what if username is not type of string?

**Review by weilaaa — APPROVED** [review]
It would be better if you could add some comments

File: pkg/authentication/identityprovider/generic/generic.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+func (g *GenericIdentity) GetUsername() string {
+	return g.Username
+}
+
+func (h HeaderProvider) Authenticate(headers map[string][]string) (identityprovider.Identity, error) {
+
+	req, err := http.NewRequest(h.Method, h.URL, nil)
+	if err != nil {
+		clog.Error("new http request err: %s", err)
```
Comment (fusida) [inline pkg/authentication/identityprovider/generic/generic.go:73]: %s not match error type

File: pkg/authentication/identityprovider/generic/generic.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+		return nil, err
+	}
+	req.Header = headers
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		clog.Error("request to generic auth error: %s", err)
```
Comment (fusida) [inline pkg/authentication/identityprovider/generic/generic.go:80]: %s not match error type

File: pkg/authentication/identityprovider/ldap/ldap.go [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+var Config = authentication.LdapConfig{}
+
+func IsLdapOpen() bool {
+	return Config.LdapIsEnable
+}
+
+type ldapProvider struct {
+	LdapObjectClass      string `yaml:"ldapObjectClass,omitempty"`
+	LdapLoginNameConfig  string `yaml:"ldapLoginNameConfig,omitempty"`
```
Comment (fusida) [inline pkg/authentication/identityprovider/ldap/ldap.go:43]: 这一行与下一行不一样，逗号后面是否有空格要统一

File: pkg/authentication/identityprovider/ldap/ldap.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+func (l *ldapIdentity) GetUsername() string {
+	return l.Username
+}
+
+func (l ldapProvider) Authenticate(username string, password string) (identityprovider.Identity, error) {
+
+	// create connection by admin account and password
+	conn, err := l.newConn()
+	if err != nil {
+		clog.Error("%s", err)
```
Comment (fusida) [inline pkg/authentication/identityprovider/ldap/ldap.go:97]: %s != error type

File: pkg/authentication/identityprovider/ldap/ldap.go [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+		BaseDN:       l.LdapBaseDN,
+		Scope:        ldap.ScopeWholeSubtree,
+		DerefAliases: ldap.NeverDerefAliases,
+		SizeLimit:    1,
+		TimeLimit:    0,
+		TypesOnly:    false,
+		Filter:       filter,
+	})
+	if err != nil {
+		clog.Error("search ldap err: %s", err)
```
Comment (fusida) [inline pkg/authentication/identityprovider/ldap/ldap.go:123]: %s != error type

File: pkg/authentication/identityprovider/ldap/ldap.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+	}
+	conn, err := ldap.Dial("tcp", host)
+	if err != nil {
+		clog.Error("connect to ldap server err: %s", err)
+		return nil, err
+	}
+
+	err = conn.Bind(l.LdapAdminUserAccount, l.LdapAdminPassword)
+	if err != nil {
+		clog.Error("bind ldap server by admin password error: %s", err)
```
Comment (fusida) [inline pkg/authentication/identityprovider/ldap/ldap.go:164]: %s != error type

File: pkg/utils/env/env.go [hunk 13]
```diff
@@ -50,12 +57,22 @@ func AuditIsEnable() bool {
 	return true
 }
 
-func AuditSVC() string {
-	r := os.Getenv("AUDIT_SVC")
+func AuditSVC() AuditSvcApi {
+	r := os.Getenv("AUDIT_URL")
+	h := os.Getenv("AUDIT_HEADER")
 	if r == "" {
```
Comment (fusida) [inline pkg/utils/env/env.go:63]: if h == "" ?

File: pkg/apiserver/middlewares/audit/audit.go [hunk 14]
```diff
@@ -141,12 +142,19 @@ func sendEvent(e *Event) {
 		return
 	}
 	buffer := bytes.NewBuffer(jsonstr)
-	request, err := http.NewRequest(http.MethodPost, "http://"+auditSvc+"/api/v1/cube/audit/cube", buffer)
+	request, err := http.NewRequest(auditSvc.Method, auditSvc.URL, buffer)
 	if err != nil {
 		clog.Error("[audit] create http request error: %s", err)
```
Comment (fusida) [inline pkg/apiserver/middlewares/audit/audit.go:147]: %s

**Review by fusida — CHANGES_REQUESTED** [review]
something need to be change

**Review by fusida — APPROVED** [review]
LGTM

File: pkg/apiserver/cubeapi/key/handle.go [hunk 15]
```diff
@@ -18,12 +18,12 @@ package key
 
 import (
 	"context"
+	"github.com/kubecube-io/kubecube/pkg/authentication/authenticators/jwt"
+	"github.com/kubecube-io/kubecube/pkg/authentication/authenticators/token"
```
Comment (Fish-pro) [inline pkg/apiserver/cubeapi/key/handle.go:22]: 导入包位置不对

Comment (zhujf1989) [reply pkg/apiserver/cubeapi/key/handle.go:22]: > The import package is in the wrong location

Thanks for review! It seems like a bug of `gofmt`. We'll use `goimport` later.

//cc please fix it manually this time @JiahuiZhao11

Comment (JiahuiZhao11) [reply pkg/apiserver/cubeapi/key/handle.go]: Have adjusted the order, thanks.

**Review by weilaaa — APPROVED** [review]
modify package import

**Review by fusida — APPROVED** [review]
LGTM

**Pull request merged**